core/src/main/scala/akka/kafka/internal/CommittingStage.scala

```diff
@@ -0,0 +1,204 @@
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.internal
+
+import java.util.regex.Pattern
+
+import akka.{Done, NotUsed}
+import akka.kafka.Subscriptions._
+import akka.kafka.scaladsl.Consumer.Control
+import akka.kafka.{ConsumerSettings, Subscription}
+import akka.stream._
+import akka.stream.scaladsl.{Flow, GraphDSL, Source}
+import akka.stream.stage._
+import org.apache.kafka.clients.consumer._
+import org.apache.kafka.clients.consumer.internals.NoOpConsumerRebalanceListener
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.unchecked.uncheckedVariance
+import scala.collection.JavaConverters._
+import scala.collection.immutable
+import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration._
+
+/**
+ * @author carsten
+ */
+final case class BidiSourceShape[+T](flowOut: Outlet[T @uncheckedVariance], flowIn: Inlet[T @uncheckedVariance], out: Outlet[T @uncheckedVariance]) extends Shape {
+  override val inlets: immutable.Seq[Inlet[_]] = flowIn :: Nil
+  override val outlets: immutable.Seq[Outlet[_]] = flowOut :: out :: Nil
+
+  override def deepCopy(): BidiSourceShape[T] = BidiSourceShape(flowOut.carbonCopy(), flowIn.carbonCopy(), out.carbonCopy())
+}
+
+object CommittingStage {
+  def apply[K, V](settings: ConsumerSettings[K, V], subscription: Subscription, flow: Flow[ConsumerRecord[K, V], ConsumerRecord[K, V], NotUsed], commitBatch: Int): Source[ConsumerRecord[K, V], Control] = {
+    val consumer = settings.createKafkaConsumer()
+    subscription match {
+      case TopicSubscription(topics) => consumer.subscribe(topics.asJava)
+      case TopicSubscriptionPattern(topics) =>
+        consumer.subscribe(Pattern.compile(topics), new NoOpConsumerRebalanceListener)
+      case Assignment(assignments) =>
+        consumer.assign(assignments.asJava)
+      case AssignmentWithOffset(assignments) =>
+        consumer.assign(assignments.keySet.asJava)
+        assignments foreach {
+          case (topic, offset) => consumer.seek(topic, offset)
+        }
+      case AssignmentOffsetsForTimes(timestampsToSearch) =>
+        val topicPartitionToOffsetAndTimestamp = consumer.offsetsForTimes(timestampsToSearch.mapValues(long2Long(_)).asJava)
+        topicPartitionToOffsetAndTimestamp.asScala.foreach {
+          case (tp, oat: OffsetAndTimestamp) =>
+            val offset = oat.offset()
+            val ts = oat.timestamp()
+            consumer.seek(tp, offset)
+        }
+    }
+    apply(consumer, flow, commitBatch)
+  }
+
+  def apply[K, V](consumer: Consumer[K, V], flow: Flow[ConsumerRecord[K, V], ConsumerRecord[K, V], NotUsed], commitBatch: Int): Source[ConsumerRecord[K, V], Control] = {
+    Source.fromGraph(GraphDSL.create(new CommittingStage(consumer, commitBatch)) { implicit b => stage =>
+      import GraphDSL.Implicits._
+
+      val handlerFlow = b.add(flow)
+
+      stage.flowOut ~> handlerFlow ~> stage.flowIn
```

Comment:
It wasn't clear to me at first why the Consumer commit and poll implementation were implemented in the `CommitStage`, but now I understand that you're building the graph internally here (or at least the segment of graph where the record is consumed, transformed, and committed).  I like this approach because it prevents the possibility of using a `committableSource` without actually committing the records, which doesn't make sense, but is currently possible

Injecting user flow in this manner also greatly simplifies the problem of keeping a reference to the Consumer before the user's flow and after.  The current implementation is quite complicated and leaves a lot to be desired IMO.